File: Homework/Objective-C/HW36/HW36/ViewController.m
//
//  ViewController.m
//  HW36
//
//  Created by Руслан on 28.11.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

#import "ViewController.h"
#import "RTDateViewController.h"

@interface ViewController () <UIPopoverPresentationControllerDelegate, UITextFieldDelegate>

@end

static NSString *kFirstName         = @"firstname";
static NSString *kLastName          = @"lastname";
static NSString *kGender            = @"gender";
static NSString *kDateOfBirth       = @"dateofbirth";
static NSString *kEducation         = @"education";

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self loadInfo];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Popovers

/*
- (void) popoverInfoForIpadWith:(UIBarButtonItem *)sender {
    
    if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
        
        UIViewController *vc = [self.storyboard instantiateViewControllerWithIdentifier:@"informationController"];
        vc.navigationItem.leftBarButtonItem = nil;
        vc.modalPresentationStyle = UIModalPresentationPopover;
        vc.preferredContentSize = CGSizeMake(CGRectGetWidth(self.view.bounds) / 2.5,
                                             CGRectGetHeight(self.view.bounds) / 4);
        
        [self presentViewController:vc animated:YES completion:nil];
        
        UIPopoverPresentationController *popover = [vc popoverPresentationController];
        popover.barButtonItem = sender;
        popover.permittedArrowDirections = UIPopoverArrowDirectionAny;
        popover.delegate = self;
    }
}

- (void) popoverDateOfBirthForIpadWith:(UITextField *)sender {
        
        UIViewController *vc = [self.storyboard instantiateViewControllerWithIdentifier:@"DateOfBirthViewController"];
        vc.modalPresentationStyle = UIModalPresentationPopover;
        
        [self presentViewController:vc animated:YES completion:nil];
        
        UIPopoverPresentationController *popover = [vc popoverPresentationController];
        popover.permittedArrowDirections = UIPopoverArrowDirectionUp;
        popover.delegate = self;
        
        CGRect popoverRect = [sender convertRect:sender.frame toView:self.view];
        
        popover.sourceView = self.view;
        popover.sourceRect = popoverRect;
}
*/

#pragma mark - Segues

- (void) prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    [self.firstNameField resignFirstResponder];
    [self.lastNameField resignFirstResponder];
    
    if ([segue.identifier isEqualToString:@"showDateOfBirth"] || [segue.identifier isEqualToString:@"showInfo"]) {
        
        RTDateViewController *vc = [segue destinationViewController];
        vc.delegate = self;
        vc.popoverPresentationController.delegate = self;
    } else if ([segue.identifier isEqualToString:@"showEducation"]) {
        
        RTEducationViewController *vc = [segue destinationViewController];
        vc.delegate = self;
        vc.popoverPresentationController.delegate = self;
    }
    
}

#pragma mark - Save and Load

- (void) saveInfo {
    
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    
    [userDefaults setObject:self.firstNameField.text forKey:kFirstName];
    [userDefaults setObject:self.lastNameField.text forKey:kLastName];
    [userDefaults setInteger:self.genderControl.selectedSegmentIndex forKey:kGender];
    [userDefaults setObject:self.dateOfBirthField.text forKey:kDateOfBirth];
    [userDefaults setObject:self.educationField.text forKey:kEducation];
    
    [userDefaults synchronize];
}

- (void) loadInfo {
    
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    
    self.firstNameField.text = [userDefaults objectForKey:kFirstName];
    self.lastNameField.text = [userDefaults objectForKey:kLastName];
    self.dateOfBirthField.text = [userDefaults objectForKey:kDateOfBirth];
    self.genderControl.selectedSegmentIndex = [userDefaults integerForKey:kGender];
    self.educationField.text = [userDefaults objectForKey:kEducation];
    
    [userDefaults synchronize];
}

#pragma mark - UITextFieldDelegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    
    if ([textField isEqual:self.firstNameField]) {
        [self.lastNameField becomeFirstResponder];
    } else {
        [self.lastNameField resignFirstResponder];
    }
    
    return YES;
}

- (BOOL)textFieldShouldBeginEditing:(UITextField *)textField {
    
    if ([textField isEqual:self.dateOfBirthField]) {
        
        [self performSegueWithIdentifier:@"showDateOfBirth" sender:textField];
        
        return NO;
    } else if ([textField isEqual:self.educationField]) {
        
        [self performSegueWithIdentifier:@"showEducation" sender:textField];
        
        return NO;
    }
    return YES;
}

#pragma mark - Actions

- (IBAction)actionSaveInfo:(UIButton *)sender {
    
    [self saveInfo];
    
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:0 inSection:1];
    [self.tableView selectRowAtIndexPath:indexPath animated:YES scrollPosition:UITableViewScrollPositionMiddle];
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
    
}

- (IBAction)actionDeleteAll:(UIButton *)sender {
    
    self.firstNameField.text = @"";
    self.lastNameField.text = @"";
    self.genderControl.selectedSegmentIndex = 0;
    self.dateOfBirthField.text = @"";
    self.educationField.text = @"";
    
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:0 inSection:2];
    [self.tableView selectRowAtIndexPath:indexPath animated:YES scrollPosition:UITableViewScrollPositionMiddle];
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - UIPopoverPresentationControllerDelegate

- (UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller traitCollection:(UITraitCollection *)traitCollection {
    return UIModalPresentationNone;
}

@end

File: Homework/Swift/HW36/HW36/ViewController.swift
//
//  ViewController.swift
//  HW36
//
//  Created by Руслан on 03.12.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

import UIKit

class ViewController:   UITableViewController, ViewControllerDelegate,
                        EducationControllerDelegate, UIPopoverPresentationControllerDelegate, UITextFieldDelegate {
    
// MARK: - Keys for NSUserDefaults
    
    let kFirstName      = "firstname"
    let kLastName       = "lastname"
    let kGender         = "gender"
    let kDateOfBirth    = "dateofbirth"
    let kEducation      = "education"

// MARK: - IBOutlet's
    
    @IBOutlet weak var firstNameField: UITextField!
    @IBOutlet weak var lastNameField: UITextField!
    @IBOutlet weak var genderControl: UISegmentedControl!
    @IBOutlet weak var dateOfBirthField: UITextField!
    @IBOutlet weak var educationField: UITextField!
    
// MARK: - Override programm functions

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.loadInfo()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

// MARK: - Actions
    
    @IBAction func actionSaveInfo(_ sender: UIButton) {
        
       self.saveInfo()
        
        let indexPath = IndexPath(row: 0, section: 1)
        self.tableView.selectRow(at: indexPath, animated: true, scrollPosition: .middle)
        self.tableView.deselectRow(at: indexPath, animated: true)
    }
    
    @IBAction func actionDeleteInfo(_ sender: UIButton) {
        
        self.firstNameField.text = "";
        self.lastNameField.text = "";
        self.genderControl.selectedSegmentIndex = 0;
        self.dateOfBirthField.text = "";
        self.educationField.text = "";
        
        let indexPath = IndexPath(row: 0, section: 2)
        self.tableView.selectRow(at: indexPath, animated: true, scrollPosition: .middle)
        self.tableView.deselectRow(at: indexPath, animated: true)
    }
    
// MARK: - Save and Load Info
    
    func saveInfo() {
        
        let userDefaults = UserDefaults.standard
        
        userDefaults.set(self.firstNameField.text, forKey: kFirstName)
        userDefaults.set(self.lastNameField.text, forKey: kLastName)
        userDefaults.set(self.dateOfBirthField.text, forKey: kDateOfBirth)
        userDefaults.set(self.educationField.text, forKey: kEducation)
        userDefaults.set(self.genderControl.selectedSegmentIndex, forKey: kGender)
        
        userDefaults.synchronize()
    }
    
    func loadInfo() {
        
        let userDefaults = UserDefaults.standard
        
        self.firstNameField.text = userDefaults.string(forKey: kFirstName)
        self.lastNameField.text = userDefaults.string(forKey: kLastName)
        self.dateOfBirthField.text = userDefaults.string(forKey: kDateOfBirth)
        self.educationField.text = userDefaults.string(forKey: kEducation)
        self.genderControl.selectedSegmentIndex = userDefaults.integer(forKey: kGender)
        
        userDefaults.synchronize()
    }
    
// MARK: - Segues
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        self.firstNameField.resignFirstResponder()
        self.lastNameField.resignFirstResponder()
        
        if segue.identifier == "showDateOfBirth" || segue.identifier == "showInfo" {
            
            let vc = segue.destination as! DateViewController
            vc.delegate = self
            vc.popoverPresentationController?.delegate = self
            
        } else if segue.identifier == "showEducation" {
            
            let vc = segue.destination as! EducationViewController
            vc.delegate = self
            vc.popoverPresentationController?.delegate = self
        }
    }
    
// MARK: - UITextFieldDelegate
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        
        if textField == self.firstNameField {
            self.lastNameField.becomeFirstResponder()
        } else {
            self.lastNameField.resignFirstResponder()
        }
        
        return true
    }
    
    func textFieldShouldBeginEditing(_ textField: UITextField) -> Bool {
        
        if textField == self.dateOfBirthField {
            
            self.performSegue(withIdentifier: "showDateOfBirth", sender: textField)
            return false
            
        } else if textField == self.educationField {
            
            self.performSegue(withIdentifier: "showEducation", sender: textField)
            return false
        }
        
        return true
    }
    
// MARK: - UITableViewDelegate
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.tableView.deselectRow(at: indexPath, animated: true)
    }
    
// MARK: - UIPopoverPresentationControllerDelegate
    
    func adaptivePresentationStyle(for controller: UIPresentationController,
                                            traitCollection: UITraitCollection) -> UIModalPresentationStyle {
        return .none
    }

}
